enable concat linear with mkldnn linear  by flag

zhuhaozhe opened the pull request:
Enable concat linear for CPU mkldnn path.
Previously, we have a concat linear in freezing passes but it not worked on CPU.
This is because `concat_linear` pattern happened after `mkldnn_weight_prepack`. And `concat_linear` only handle `addmm/mm` etc.

```
addmm -> mkldnn linear
addmm -> mkldnn linear -> cannot concat

# only worked when disable mkldnn
addmm ->
addmm -> concat linear
```
Now we changed `mkldnn linear` related pass numbers larger than `concat_linear` pass numbers.

```
addmm -> concat linear -> mkldnn linear
addmm ->

```
So it can work fine with mkldnn linear now.

Also, since concat linear not always have benefits. We add 1 flag `config.cpp.enable_concat_linear` and set default value to False. User can enable this by their need. 





Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139048



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

jgong5 commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -930,6 +930,14 @@ def _is_packable_linear(match):
         Check if the node is supported for MKLDNN linear.
         """
+
+        def is_constant_weight(weight):
```
Can you revise the naming of the function? It does not only check the weight is constant, right?

zhuhaozhe replied:
Yes, thanks. Revised to is_const_or_cat_by_const

jgong5 commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -748,7 +748,7 @@ def _recover_linear():
                 KeywordArg("reshape_2"),
             ),
-            pass_number=1,
+            pass_number=2,
```
Explain the change?

zhuhaozhe replied:
Previously `mkldnn_linear` pass is 0 and `_recover_linear` is based on `mkldnn_linear` so it should be 1.
Now we set `mkldnn_linear` pass to 1 (since concat linear is 0), then the `recover_linear` pass number should be 2.

jgong5 replied:
Please add the note in the code too.

zhuhaozhe replied:
Thanks for advice, added this note in code.

jgong5 commented on torch/_inductor/config.py:
```diff
@@ -879,6 +879,9 @@ class cpp:
     # Whether to enable masked vectorization for the tail_loop.
     enable_loop_tail_vec = True
 
+    # Whether to enable concat linear for cpu device
```
Can you also comment on why it defaults to `False`?

zhuhaozhe replied:
Yes, thanks for advice, added.

jgong5 reviewed (changes requested):
Could you please update the PR description on changes?

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot merge

zhuhaozhe commented:
@pytorchbot  rebase

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable concat linear with mkldnn linear  by flag"

Enable concat linear for CPU mkldnn path.
Previously, we have a concat linear in freezing passes but it not worked on CPU.
This is because `concat_linear` pattern happened after `mkldnn_weight_prepack`. And `concat_linear` only handle `addmm/mm` etc.

```
addmm -> mkldnn linear
addmm -> mkldnn linear -> cannot concat

# only worked when disable mkldnn
addmm ->
addmm -> concat linear
```
Now we changed `mkldnn linear` related pass numbers larger than `concat_linear` pass numbers.

```
addmm -> concat linear -> mkldnn linear
addmm ->

```
So it can work fine with mkldnn linear now.

Also, since concat linear not always have benefits. We add 1 flag `config.cpp.enable_concat_linear` and set default value to False. User can enable this by their need. 








cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]
